src/scopes.js

```diff
@@ -56,3 +60,59 @@ exports.scopeMatch = function(scopePatterns, scopesets) {
     });
   });
 };
+
+/**
+ * Finds scope intersections between two scope sets.
+ */
+exports.scopeIntersection = (scopeset1, scopeset2) => {
+  validateScopePatterns(scopeset2);
+  validateScopePatterns(scopeset1);
+
+  const junctions = [];
+  const scope1 = scopeset2.filter(s => s);
+  const scope2 = scopeset1.filter(s => s);
+
+  // case where scope1 or scope2 is a wildcard
+  if (scope1.length === 1 && scope1[0] === '*') {
+    return scope2;
+  } else if (scope2.length === 1 && scope2[0] === '*') {
+    return scope1;
+  }
+
+  scope2.forEach((p) => {
+    scope1.forEach((s) => {
+      const s1 = s.split(':');
+      const s2 = p.split(':');
```

Comment:
Hm, `:` isn't special in scopes..

Let's go back to first principles here: given two scopes S1 and S2, either
 * S1 == S2
 * [S1] satisfies [S2], meaning that S1 ends with `*` and S1 and S2 are equal up to the position of that `*`
 * [S2] satisfies [S1], meaning that S2 ends with `*` and S1 and S2 are equal up to the position of that `*`
 * neither scope satisfies the other

In the first case, the intersection is [S1]; in the second, [S2]; in the third, [S1], and in the fourth, [].

The intersection of two sets of scopes is the union of the all pairwise intersections.